roles/testnode/files/libexec/smart.sh

```diff
@@ -0,0 +1,284 @@
+    if [ "$areca_serial" = "$smart_serial" ]
+    then
+      echo "($dl)"
+    fi
+  done
+}
+
+# Tally missing and failed drives connected to Areca RAID
+areca_failed ()
+{
+  # Store output of cli64 to reduce repeated executions
+  cli64out=$(sudo cli64 disk info | grep -E "Slot#[[:digit:]]")
+  # Missing (N.A.) drives
+  for drive in $(echo "$cli64out" | grep -E "Slot#[[:digit:]]" \
+  | grep "N.A." | awk '{ print $1 }')
+  do
+    messages+=("Drive $drive is missing")
+    let "failingdrives+=1"
+    rc=2
+  done
+  # Hard failed drives
+  for drive in $(echo "$cli64out" | grep -E "Slot#[[:digit:]]" \
+  | grep 'Failed' | awk '{ print $1 }')
+  do
+    messages+=("Drive $drive failed")
+    let "failingdrives+=1"
+    rc=2
+  done
+}
+
+# Standard SATA/SAS drive smartctl check
+normal_smart ()
+{
+  # The grep regex will include drives named sdaa, for example
+  numdrives=$(cat /proc/partitions | grep -w 'sd[a-z]\|sd[a-z]\{2\}' | wc -l)
+  for l in $(cat /proc/partitions | grep -w 'sd[a-z]\|sd[a-z]\{2\}' \
+  | awk '{ print $NF }')
+  do
+    failed=false
+    output=$(sudo smartctl -a /dev/$l | grep -E "^  "5"|^"197"|^"198"" \
+    | awk '{ print $NF }' | tr '\n' ' ')
+    outputcount=$(echo $output | wc -w)
+    # Check if drive is SSD and set var accordingly
+    if sudo smartctl -i /dev/$l | grep -q 'Solid State Device'; then
+      is_ssd=true
+    else
+      is_ssd=false
+    fi
+    # Only continue if we received 3 SMART data points and drive is not SSD
+    if [ "$outputcount" = "3" ] && [ "$is_ssd" = false ]
+    then
+      read reallocated pending uncorrect <<< $output
+      if [ "$reallocated" != "0" ]
+      then
+        messages+=("Drive $l has $reallocated reallocated sectors")
+        failed=true
+        # A small number of reallocated sectors is OK
+        if [ "$reallocated" -le 5 ]
+        then
+          rc=1 # Warn if <= 5
+        else
+          rc=2 # Crit if >5
+        fi
+      fi
+      if [ "$pending" != "0" ]
+      then
+        messages+=("Drive $l has $pending pending sectors")
+        failed=true
+        rc=2
+      fi
+      if [ "$uncorrect" != "0" ]
+      then
+        messages+=("Drive $l has $uncorrect uncorrect sectors")
+        failed=true
+        rc=2
+      fi
+    elif [ "$outputcount" != "3" ] && [ "$is_ssd" = false ]
+    then
+      messages+=("Drive $l returned $outputcount of 3 expected attributes")
+      unknownmsg="SMART data could not be read for one or more drives"
+      rc=3
+    # Set no return code and assume any SSD is healthy for now
+    elif [ "$is_ssd" = true ]
+    then
+      messages+=("Drive $l is an SSD.  Not yet supported.")
+    else
+      messages+=("Error processing data for drive $l")
+      rc=3
+    fi
+    # Make sure drives with multiple types of bad sectors only get counted once
+    if [ "$failed" = true ]
+    then
+      let "failingdrives+=1"
+    fi
+  done
+}
+
+## Call main() function
+main
+
```

Comment:
it's kind of weird to have a "main" function that's not "everything in the script that's not a function" (as main traditionally is in C, Python, etc.)  I'd either subsume the rest of the code into main, or rename main.